### ISSUE TITLE
Additional scale & shift loss

### ISSUE THREAD

**Issue by cdcseacave** [zju3dv/Murre#12]
Thank you for sharing this project, it works reasonable well in most scenes I tested.

However I noticed that even if the generated depth-maps are bootstrapped to the SfM point cloud, their scale and shift still vary from 1 & 0, across datasets and across the images in a dataset. Do you think it would help if you add an explicit loss function during training that penalize the scale and shift deviation of the generated depth-map?

I implemented a depth-map scale and shift aligning script to SfM in OpenMVS, and this is the output on the 37 image buildings dataset from UrbanScene3D:
```
(murre) ubuntu@150-136-33-157:~/data/BuildingsDrone$ python3 /home/ubuntu/Murre/ImportDMAPs.py --scene scene.mvs --input murre/depth_npy/ --ext _pred.npy -r -v
Scene scene.mvs loaded: 37 images

RANSAC stats: 6 iterations, 1.3778 score, 2073 / 2204 inliers
Optimization stats: 15 iterations, 1.1666 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 0.9937, shift: -0.4789

RANSAC stats: 4 iterations, 1.1117 score, 2288 / 2337 inliers
Optimization stats: 18 iterations, 1.0811 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0096, shift: -2.9383

RANSAC stats: 4 iterations, 1.4103 score, 2233 / 2276 inliers
Optimization stats: 20 iterations, 1.2336 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0102, shift: -3.0473

RANSAC stats: 4 iterations, 3.6728 score, 2031 / 2083 inliers
Optimization stats: 21 iterations, 1.2298 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0063, shift: -2.4642

RANSAC stats: 6 iterations, 10.6718 score, 1657 / 1767 inliers
Optimization stats: 24 iterations, 0.9747 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 0.9987, shift: -1.3517

RANSAC stats: 5 iterations, 5.7416 score, 2162 / 2264 inliers
Optimization stats: 23 iterations, 1.0019 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0096, shift: -3.1597

RANSAC stats: 5 iterations, 1.0867 score, 2243 / 2331 inliers
Optimization stats: 15 iterations, 0.9662 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0113, shift: -3.2552

RANSAC stats: 4 iterations, 1.2112 score, 2418 / 2483 inliers
Optimization stats: 22 iterations, 0.9225 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0071, shift: -2.6863

RANSAC stats: 6 iterations, 2.9216 score, 2481 / 2614 inliers
Optimization stats: 19 iterations, 1.6220 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 0.9983, shift: -0.7349

RANSAC stats: 5 iterations, 1.6986 score, 2339 / 2425 inliers
Optimization stats: 17 iterations, 1.2000 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0041, shift: -1.6268

RANSAC stats: 7 iterations, 1.3705 score, 1925 / 1999 inliers
Optimization stats: 3 iterations, 1.4142 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 0.9990, shift: -1.2417

RANSAC stats: 5 iterations, 3.2283 score, 5063 / 5333 inliers
Optimization stats: 16 iterations, 1.2937 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0036, shift: -1.6828

RANSAC stats: 5 iterations, 1.3679 score, 1747 / 1834 inliers
Optimization stats: 19 iterations, 1.4336 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0050, shift: -2.1089

RANSAC stats: 5 iterations, 2.3435 score, 4855 / 5093 inliers
Optimization stats: 18 iterations, 1.3553 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0127, shift: -3.0518

RANSAC stats: 6 iterations, 1.4383 score, 1935 / 2047 inliers
Optimization stats: 21 iterations, 1.5851 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0016, shift: -1.5775

RANSAC stats: 5 iterations, 1.3024 score, 3483 / 3601 inliers
Optimization stats: 21 iterations, 1.1422 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0110, shift: -3.1749

RANSAC stats: 7 iterations, 1.8342 score, 2105 / 2312 inliers
Optimization stats: 21 iterations, 1.4178 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 0.9955, shift: -0.5564

RANSAC stats: 5 iterations, 1.0340 score, 2899 / 2986 inliers
Optimization stats: 19 iterations, 0.9415 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0106, shift: -3.1020

RANSAC stats: 6 iterations, 1.3155 score, 2615 / 2793 inliers
Optimization stats: 22 iterations, 1.3096 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0039, shift: -2.0997

RANSAC stats: 5 iterations, 3.4919 score, 2845 / 2965 inliers
Optimization stats: 21 iterations, 1.1190 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0166, shift: -3.9548

RANSAC stats: 5 iterations, 1.3316 score, 2270 / 2381 inliers
Optimization stats: 20 iterations, 1.3024 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0068, shift: -2.4557

RANSAC stats: 6 iterations, 10.0132 score, 3203 / 3407 inliers
Optimization stats: 22 iterations, 1.1435 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0085, shift: -2.5188

RANSAC stats: 6 iterations, 5.3132 score, 3178 / 3328 inliers
Optimization stats: 19 iterations, 1.2901 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0083, shift: -2.8382

RANSAC stats: 5 iterations, 1.2017 score, 2291 / 2356 inliers
Optimization stats: 21 iterations, 1.0158 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0063, shift: -2.2540

RANSAC stats: 6 iterations, 1.4916 score, 4301 / 4650 inliers
Optimization stats: 20 iterations, 1.3692 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0090, shift: -2.5810

RANSAC stats: 6 iterations, 4.6794 score, 3163 / 3380 inliers
Optimization stats: 19 iterations, 1.1661 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0127, shift: -3.4387

RANSAC stats: 5 iterations, 15.6461 score, 2568 / 2647 inliers
Optimization stats: 17 iterations, 1.4307 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0005, shift: -1.0863

RANSAC stats: 6 iterations, 2.0414 score, 4947 / 5316 inliers
Optimization stats: 16 iterations, 1.4341 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0075, shift: -2.2958

RANSAC stats: 6 iterations, 31.0949 score, 2751 / 2903 inliers
Optimization stats: 26 iterations, 1.1708 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0050, shift: -2.3448

RANSAC stats: 5 iterations, 2.4441 score, 2120 / 2193 inliers
Optimization stats: 21 iterations, 1.4677 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0020, shift: -1.6079

RANSAC stats: 5 iterations, 1.9140 score, 2231 / 2311 inliers
Optimization stats: 21 iterations, 0.7981 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0066, shift: -2.3538

RANSAC stats: 5 iterations, 9.7839 score, 2235 / 2320 inliers
Optimization stats: 21 iterations, 1.0229 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 0.9985, shift: -1.2732

RANSAC stats: 4 iterations, 17.1312 score, 2166 / 2224 inliers
Optimization stats: 24 iterations, 0.6739 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0001, shift: -1.5155

RANSAC stats: 4 iterations, 13.4115 score, 2244 / 2269 inliers
Optimization stats: 21 iterations, 0.7114 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0081, shift: -2.9069

RANSAC stats: 4 iterations, 13.7855 score, 2218 / 2281 inliers
Optimization stats: 25 iterations, 0.7727 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0000, shift: -1.6058

RANSAC stats: 4 iterations, 1.7613 score, 2417 / 2460 inliers
Optimization stats: 16 iterations, 0.9735 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0098, shift: -3.0935

RANSAC stats: 5 iterations, 1.3728 score, 2407 / 2500 inliers
Optimization stats: 18 iterations, 0.8775 score, CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
Estimated scale: 1.0143, shift: -3.8258
```

**Comment by 3i-cao** [comment]
@cdcseacave I tested with sparse data output of colmap as an input of Murre and it still has a few bugs in generate depth, can you help me to provide an example of data you test successfully to help me debug it. Thanks

**Comment by Parskatt** [comment]
I'm wondering if this might rather be due to the RANSAC implementation, as there's no threshold provided to sklearn. I don't believe that the shift should be this large.

**Comment by ghy0324** [comment]
@cdcseacave Thanks very much for your interest in our work! The depth alignment shifts may stem from the following factors:

1. The VAE itself introduces a small but measurable bias. Passing GT depth through the encoder–decoder leads to an average error of approximately 0.5% (see [Notebook 1](https://ghy0324.github.io/project_page_assets/murre/test_vae_error.html)). If higher precision is required, we may need to improve the VAE or adopt a different paradigm.

2. SfM depth estimates contain errors, and there is a shift when aligning the raw GT depth with the SfM depth (see the end of [Notebook 1](https://ghy0324.github.io/project_page_assets/murre/test_vae_error.html)).

3. A more significant reason is that the 3D Ken Burns dataset does not provide GT camera poses (see: https://github.com/sniklaus/3d-ken-burns/issues/71), so we recovered the camera poses with SfM. The resulting sparse depth and the GT depth are on different scales, so we first aligned them (only scale, no shift): for each scene we divided the GT depth by the sparse depth pixel‑wise, took the median of that ratio as a scale factor, and multiplied the sparse depth by this value. [Notebook 2](https://ghy0324.github.io/project_page_assets/murre/3d-ken-burns-scale-ratio.html) gives the details; `scale_dict` is the final value we used in experiments, and the histograms there show that this procedure is imperfect, but so far we have not found a better solution. Maybe we should switch to other dataset.

Empirically, we observe alignment shifts in both indoor and outdoor scenes, with the outdoor scenes being more affected. This suggests that all three factors play a role, but factor 3 likely dominates (note that our training used only Hypersim and 3D Ken Burns; Hypersim has GT poses and contains only indoor scenes).

**Comment by PangChen0402** [comment]
@ghy0324 Hi, Thank you for sharing! When I read depthanything v2. they said

![Image](https://github.com/user-attachments/assets/3e26b4a6-8f0d-4112-a703-1ae9f212c3be)

![Image](https://github.com/user-attachments/assets/92c222b1-2dea-4d2b-bb9a-5c022f225c00)

Have you considered using synthetic images for tranning?
